src/components/forms/CharacterCount/CharacterCount.tsx

```diff
@@ -0,0 +1,165 @@
+import classnames from 'classnames'
+
+import { TextInput, TextInputRef } from '../TextInput/TextInput'
+import { Textarea, TextareaRef } from '../Textarea/Textarea'
+import { FormGroup } from '../FormGroup/FormGroup'
+
+interface CharacterCountProps {
+  id: string
+  name: string
+  maxLength: number
+  className?: string
+  label?: React.ReactNode
+  isTextArea?: boolean
+  textRef?: // currently getting an error when adding this as ref
+  | string
+    | ((instance: HTMLInputElement | null) => void)
+    | ((instance: HTMLTextAreaElement | null) => void)
+    | React.RefObject<HTMLInputElement>
+    | React.RefObject<HTMLTextAreaElement>
+    | null
+    | undefined
+}
+
+export const CharacterCount = (
+  props: CharacterCountProps &
+    React.InputHTMLAttributes<HTMLInputElement> &
+    React.TextareaHTMLAttributes<HTMLTextAreaElement>
+): React.ReactElement => {
+  const {
+    id,
+    name,
+    className,
+    label,
+    maxLength,
+    defaultValue,
+    isTextArea,
+    textRef,
+    type, // Conflicts with type attribute already set on TextInput component
+    ...inputProps
+  } = props
+
+  const defaultLength =
+    defaultValue === undefined ? 0 : defaultValue.toString().length
+  const maxNum = maxLength === undefined ? 0 : maxLength
+
+  /* Ideally defined as i18n translation strings */
+  const emptyMessageFormat = `${maxLength} characters allowed`
+  const remainingPluralFormat = '$0 characters left'
+  const remainingSingularFormat = '$0 character left'
+  const overSingularFormat = '$0 character over limit'
+  const overPluralFormat = '$0 characters over limit'
+
+  const limitMessage = (len: number): string => {
+    let limitMessage
+    if (len === 0) {
+      return emptyMessageFormat
+    } else if (len <= maxNum) {
+      switch (maxNum - len) {
+        case 1:
+          return remainingSingularFormat.replace('$0', '1')
+        default:
+          return remainingPluralFormat.replace('$0', (maxNum - len).toString())
+      }
+    } else {
+      switch (len - maxNum) {
+        case 1:
+          return overSingularFormat.replace('$0', '1')
+        default:
+          return overPluralFormat.replace('$0', (len - maxNum).toString())
+      }
+    }
+  }
+
+  const [state, setState] = useState({
+    characterLength: defaultLength,
+    limitMessage: limitMessage(defaultLength),
+    invalidMessage: defaultLength > maxNum,
+  })
+
+  const classes = classnames('usa-character-count__field', className)
+  const messageClasses = classnames(
+    'usa-hint',
+    'usa-character-count__message',
+    { 'usa-character-count__message--invalid': state.invalidMessage }
+  )
+
+  const messageId = id + '-info'
+
+  const handleChange = (
+    e: React.ChangeEvent<HTMLInputElement> &
+      React.ChangeEvent<HTMLTextAreaElement>
+  ): void => {
+    const len = e.target.value.length
+    setState({
+      characterLength: len,
+      limitMessage: limitMessage(len),
+      invalidMessage: len > maxLength,
+    })
```

Comment:
A few nitpicks here:
- Is there a need to track the `characterLength` in state? 
- Suggestion -  rename `invalidMessage` to make it clearer, maybe`isOverLimit`
- Suggestion -  spread state first to prevent bugs when making all of state is  one object
``` 
setState((state) => ({
      ...state,
      limitMessage: limitMessage(len),
      invalidMessage: len > maxLength,
    }))
```

Reply:
Agreed, good points! Yep I did read about setState not being a merge operation in this case.